### ISSUE TITLE
Layout save/restore

### ISSUE THREAD

**Issue by SirCmpwn** [SirCmpwn/sway#1005]
Layout save/restore is an i3 feature that serializes your layout to JSON and attempts to arrange windows in the same way in a later session. First step is to research how i3 does it and propose some strategies for implementing it in sway.

**Comment by cyphar** [comment]
I did a bit of research [using the documentation as a reference](https://i3wm.org/docs/layout-saving.html). Effectively the way this is implemented (I don't use this feature at all) is this:

* When generating the `i3-save-tree` output it includes all of the obvious information about container placement and so on -- which for us would just be a matter of recursively walking down the tree.

* However, it includes "swallowing" information which is effectively a regular expression that matches `WM_CLASS` (which is a tuple of (instance, class)), `_NET_WM_NAME` (title) and `WM_WINDOW_ROLE` (window role). When a program is started in an i3 session and it matches all of the "swallowing" information, it will place that program in the loaded pane.

* Note that the output of `i3-save-tree` is not actually standard-conformant JSON. IMO we should not follow their lead, and we should always output standard compliant JSON.

So from what I can see there are a few things we need to implement and/or consider:

0. Is there a nicer way of implementing "swallowing" in Wayland than in X? The above seems quite hacky to me.
1. What is the analogy for `WM_CLASS`, `_NET_WM_NAME` and ` WM_WINDOW_ROLE` in the Wayland (and wlc) protocols?
2.

**Comment by SirCmpwn** [comment]
I've decided that we are just not going to support this i3 feature, ever.

**Comment by mahmoudhossam** [comment]
@SirCmpwn Any specific reason why?

**Comment by SirCmpwn** [comment]
It's too complicated and hacky for too little benefit. I don't like the design constraints it imposes upon sway.

**Comment by hyiltiz** [comment]
Would you reconsider your decision? This feature is actually quite useful in some (could even say rare) professional settings, e.g. setting up a music production with a midi mixer, synthesizer, sequencer, and composition programs open in their specific layout every time, or just general software development with editors, compilers, debuggers and test windows. 

While there is no point in being i3 compliant on this, I think a good discussion in trying to come up with a better solution would be quite beneficial.

**Comment by SirCmpwn** [comment]
No, I won't reconsider. However, you should be able to accomplish something similar with a script that starts up the software you need and arranges it appropriately via IPC.

**Comment by hyiltiz** [comment]
I would love to say "Would you like to reconsider whether you would like to reconsider 'No, I won't reconsider' ", but that might easily overflow the buffer if not implemented correctly. Anyway, I do think writing a script for that should be straightforward so thanks!

**Comment by flux242** [comment]
it's a shame that this great i3 feature is dropped simply because "I'm too lazy to implement something I personally do not use". Please tell me how am I suppose to do swallowing in a script? How do I assign a program window to a specific layout window using IPC? If I start a program I only know its process id and what is available in the /proc/proc_id.

**Comment by SirCmpwn** [comment]
>It's too complicated and hacky for too little benefit. I don't like the design constraints it imposes upon sway.

This is why it was dropped, not because, quote, "I'm too lazy to implement something I personally do not use." You can use IPC to move a window to a specific workspace over IPC with criteria. You can split the window and move things around, too. The pid associated with a window is included in the IPC_GET_TREE output, along with a unique ID you can use for criteria.

**Comment by flux242** [comment]
give an IPC command example please of moving a program window into a specific layout window on a workspaceX

**Comment by flux242** [comment]
right, why there can't be a method that does exactly this if it is so easy? It would just have to parse a layout file additionally, right? Now using your approach what would I do if I wanted to change my layout a bit, like resize here and there provided that my layout is a little bit more complex rather that just 3 windows? Currently I just have to resize and overwrite my layout file. And btw I'm inside of a bash script so how do I get an IPC notification?

**Comment by SirCmpwn** [comment]
>right, why there can't be a method that does exactly this if it is so easy?

See

>It's too complicated and hacky for too little benefit. I don't like the design constraints it imposes upon sway.

If you want to write a script which implements i3 layout save/restore as an external program, you should do so, but no one is going to write it for you.

>And btw I'm inside of a bash script so how do I get an IPC notification?

By not being in a bash script

**Comment by flux242** [comment]
nope, I'm not going to implement a script because
> It's too complicated and hacky

without proper support from underlying framework. Your script "example" could only work for very simple layouts. It's because if you have a generic layout description you'd have to start with a specific program window traversing the layout up the tree trying to figure out where that window would land in the end and how the inverted IPC command chain would look like. 

btw feature that allows restoring a complex working environment (up to starting terminals with specific commands) just by a keyboard shortcut brings rather huge benefit than "too little benefit". Lacking such a feature is a road block for me => no switching from i3. Chusikowski

**Comment by Streetwalrus** [comment]
I agree that i3's implementation of this feature has its own set of shortcomings. However, the proposed alternative of using IPC feels rather clunky since all it allows is akin to replaying a keyboard macro.

A more powerful alternative would be to extend the API to allow passing a tree of *existing* windows referenced by ID or criteria, with optional hints for layout and sizing, and have sway apply that atomically.
This would feel much cleaner than i3's swallow windows (which can only capture windows that do not yet exist), while providing equivalent, if not more flexible functionality.

**Comment by SirCmpwn** [comment]
I would consider a patch, but I'm not interested in doing this myself and I'm worried about the complexity/value tradeoff.

**Comment by Streetwalrus** [comment]
That's all I needed to hear. I would gladly write this feature myself, but I'm currently focusing on other projects so i3 is still my daily driver for now. I'll probably come back to it later, unless someone else is interested in doing it before me.

**Comment by JonnyHaystack** [comment]
> I agree that i3's implementation of this feature has its own set of shortcomings. However, the proposed alternative of using IPC feels rather clunky since all it allows is akin to replaying a keyboard macro.
> 
> A more powerful alternative would be to extend the API to allow passing a tree of _existing_ windows referenced by ID or criteria, with optional hints for layout and sizing, and have sway apply that atomically.
> This would feel much cleaner than i3's swallow windows (which can only capture windows that do not yet exist), while providing equivalent, if not more flexible functionality.
> 
> Note that my proposal differs from i3 in two ways:
> 
> * i3 expects a path to a JSON file containing the layout. This is dumb, the entire layout should be passed over IPC instead. Clients may choose to read a layout from a file, but that's up to them.
> * When you call `append_layout`, i3 spawns placeholder windows, and stores the criteria in the container tree. When a new window is mapped and i3 attempts to manage it, it searches the tree for a matching placeholder and substitutes the first positive match. This requires special handling in various components of the window manager, and can't operate on existing windows.
>   My proposal is rather self-contained: since it should operate on existing windows, everything can be done immediately in the command handler, without having to spawn placeholders and/or store additional metadata in the tree.

I completely agree with all of this. The placeholder system in i3 is not very good and you have to use hacks like unmapping and remapping windows with xdotool to get a layout to apply to existing windows.

Let me know when you get time to start working on it. I'd be eager to help out.

**Comment by JonnyHaystack** [comment]
I've recently been looking at the code that i3 uses when restoring a layout.

```
json_content_t content = json_determine_content(buf, len);
LOG("JSON content = %d\n", content);
if (content == JSON_CONTENT_UNKNOWN) {
    ELOG("Could not determine the contents of \"%s\", not loading.\n", path);
    yerror("Could not determine the contents of \"%s\".", path);
    goto out;
}

Con *parent = focused;
if (content == JSON_CONTENT_WORKSPACE) {
    parent = output_get_content(con_get_output(parent));
} else {
    /* We need to append the layout to a split container, since a leaf
     * container must not have any children (by definition).
     * Note that we explicitly check for workspaces, since they are okay for
     * this purpose, but con_accepts_window() returns false for workspaces. */
    while (parent->type != CT_WORKSPACE && !con_accepts_window(parent))
        parent = parent->parent;
}
DLOG("Appending to parent=%p instead of focused=%p\n", parent, focused);
char *errormsg = NULL;
tree_append_json(parent, buf, len, &errormsg);
if (errormsg != NULL) {
    yerror(errormsg);
    free(errormsg);
    /* Note that we continue executing since tree_append_json() has
     * side-effects — user-provided layouts can be partly valid, partly
     * invalid, leading to half of the placeholder containers being
     * created. */
} else {
    ysuccess(true);
}
```


So it checks for the type of the data it loads from the file whose path you pass
to `append_layout`, and defaults to con if it doesn't find any "type"
properties.

If the content is of type "workspace", it appends the layout to the currently
focused output's workspace array. If it is a con, it searches upwards from the
currently focused node until it finds a workspace, then it appends the layout to
the workspace's child array.

The fact that it only appends nodes to an array and does not modify or replace existing nodes means that it is actually quite clean. I've come to appreciate the simplicity of this approach, as opposed to rewriting the existing tree.

The only part of their implementation that still seems bad to me is that placeholder container matching is only done when windows are created. We could just implement layout restoring almost exactly the same as i3, but just search the whole workspace for matching windows right after the layout is appended, instead of when new windows are mapped.

Or we could have the best of both worlds and check for matches after layout appending, and on window creation, for those use cases where placeholder windows/deferred swallowing is desirable (these cases really do exist). The caveat of this is obviously that it's less self-contained because of having to monitor window creation events.

Either way it doesn't seem like this would be that hard to implement.

**Comment by flux242** [comment]
> The only part of their implementation that still seems bad to me is that placeholder container matching is only done when windows are created. We could just implement layout restoring almost exactly the same as i3, but just search the whole workspace for matching windows right after the layout is appended, instead of when new windows are mapped.
lets say I have 100 windows in my layout. Right now I have a simple script that loads specific layout and then starts 100 programs **in the background**. Your suggestion would imply that I have to implement additional 'wait_for_all_100_windows_be_opened_before_loading_layout' and which is more important 'do_x_if_window_y_wont_appear_within_z_sec' logics. Is that what you're saying  @JonnyHaystack ?

**Comment by norcalli** [comment]
@ddevault your assertion that this could be easily done in a script is incorrect. I know because I tried it. sway doesn't have good enough commands to make it easy to be accomplished for an arbitrary layout. For example, there is no "move to this container." Instead you have to make a mark, and then move to that mark. This ends up being error prone. Another thing I tried is creating a placeholder window so that I knew its con_id and I could target it without needing to use marks as a place to drop windows on. This too is error prone as the timing required to make sure that windows transfer is not so straightforward.

Finally, there's also not a good way of modifying the layout of windows without focusing them. You may thing that criteria targetting would work, but it doesn't work reliably. And this is all from 2020 when I tried it. At the time that you made the suggestion to make a script in 2017, I am certain sway was in an even less adequate state to achieve it.

Maybe I'm not smart enough to get it done, but then again that also proves the value of having it built into sway or having an official solution.

I could try to spend more time on it, but I have already sunk 3 hours into it. It's likely not worth it to do in such a hacky way. And if you want dozens of other people to spend/waste their time simply because you are still under the misconception that this is "easy to do" then I think it would be fair to apologize for the time that others have spent and will spend time on this.

**Comment by emersion** [comment]
Ref https://github.com/swaywm/sway/pull/3022

**Comment by Hubro** [comment]
Are there any good options for restoring layouts at this point? I'm a software developer and I use `append_layout` daily to start my work environments, as well as position my chat windows (I use like 8 different chat clients.) The lack of this functionality is my main blocker with Sway at the moment.

If anybody has any good workarounds using IPC I would love to see some examples. I'm not savvy enough with Sway to have any clue what ddevault is talking about when he says things like `IPC_GET_TREE`.

**Comment by yellowhat** [comment]
Hi,
I managed to use `i3ipc-python`, to open a multiple windows at the same time in a certain order:
```python
#!/usr/bin/env python3

from i3ipc import Connection
from time import sleep

conn = Connection()

def run(cmd):
    ilen = len(conn.get_tree().workspaces()[-1].descendants())
    conn.command(cmd)
    sleep(0.2)
    while len(conn.get_tree().workspaces()[-1].descendants()) == ilen:
        True
    return max(i.id for i in conn.get_tree().workspaces()[-1].descendants())

conn.command("workspace file")
con_id1 = run("exec wayst -e sf /")
con_id2 = run("exec wayst -e sf ~/Downloads")
con_id3 = run(f"[con_id={con_id1}] focus; splitv; exec wayst -e sf /tmp")
con_id4 = run(f"[con_id={con_id2}] focus; splitv; exec wayst -e sf /mnt")
conn.command(f"[con_id={con_id1}] focus")
```
Not a fan of the `sleep` command, any suggestions are welcome

**Comment by Hubro** [comment]
> Hi,
> I managed to use `i3ipc-python` to open a multiple windows at the same time in a certain order:
> 
> ```python
> #!/usr/bin/env python3
> 
> from i3ipc import Connection
> 
> conn = Connection()
> 
> def run(cmd):
>     conn.command(cmd)
>     ilen = len(conn.get_tree().workspaces()[-1].descendants())
>     while len(conn.get_tree().workspaces()[-1].descendants()) == ilen:
>         True
>     con_id = max(i.id for i in conn.get_tree().workspaces()[-1].descendants())
>     return con_id
> 
> conn.command("workspace file")
> con_id1 = run("exec wayst -e sf /")
> con_id2 = run("exec wayst -e sf ~/Downloads")
> con_id3 = run(f"[con_id={con_id1}] focus; splitv; exec wayst -e sf /tmp")
> con_id4 = run(f"[con_id={con_id2}] focus; splitv; exec wayst -e sf /mnt")
> conn.command(f"[con_id={con_id1}] focus")
> ```
> 
> Any suggestions are welcome

Aha, I see... Your example would probably work in some simple cases, but I can already see it falling apart with applications like Discord, where it displays a loading splash screen for a little while before starting, and often also starts an updater. Although this could be worked around by waiting for a new window that matches certain filters.

Also it has to open applications one at a time, which could be potentially very slow, but getting around that would be pretty complicated :thinking: I'm imagining a script that would launch a bunch of applications simultaneously, then loop on `get_tree` and resolve each new window ID based on a filter and set each new window to floating, then a different thread would execute an ordered list of movement commands to place the windows in the right position after "unfloating" them. The script could be made reusable by a simple DSL, something like this:

```python
from sway_layout import Row, Column, Application, launch

launch(
    Column(
        Application("chromium --app=https://messenger.com", cls="Chromium",  instance="^Messenger$"),
        Application("slack", cls="Slack",  instance="^Messenger$"),
    ),
    Column(
        Application("viber", cls="^Viber$"),
        Row(
            Application("discord", cls="^discord$", instance="^discord$", window_type="normal", name="- Discord$"),
            Application("teamspeak3", cls="^TeamSpeak3$"),
        )
    )
)
```

I don't see any reason this wouldn't be possible to implement. I will probably take a swing at it when I get some free time, I am really missing this feature.

**Comment by 9ary** [comment]
This is my take on the problem, it's definitely not bulletproof but it works most of the time for me. https://github.com/9ary/dotfiles/blob/master/i3/ws-1.py

**Comment by tgunnoe** [comment]
@9ary can you explain a bit to me how this works? I couldn't get your example working after a day or so, but it could be due to my lack of experience with the sway IPC.

in particular, https://github.com/9ary/dotfiles/blob/master/i3/ws-1.py#L83
```
event_loop = asyncio.create_task(sway.main())
```
It seems to never respond, or report back anything after it hits this point. What am I doing wrong?

**Comment by 9ary** [comment]
Yeah, it blocks until every leaf node has a match, so it won't do anything until all members of your layout exist (I would have to implement placeholders otherwise, which is significantly more work, and my workflow doesn't require them).

**Comment by tgunnoe** [comment]
Thanks for the explanation. Got it working after learning more about the script and `get_tree`

**Comment by JonnyHaystack** [comment]
> 
> 
> This is my take on the problem, it's definitely not bulletproof but it works most of the time for me. https://github.com/9ary/dotfiles/blob/master/i3/ws-1.py
> Again, the ability to pass a tree into sway and have it arrange windows accordingly would help a lot for writing external tools to do this. It doesn't even need to be smart and match criteria, just identify windows by their ID and leave the rest up to the application.

@9ary How would you match the window by ID considering the ID the next time you launch the program will be different from when you saved the layout?

**Comment by 9ary** [comment]
The point is that an external program could query the tree to identify the desired windows, and even spawn placeholder windows, effectively taking the complexity out of sway, and enabling more complex use cases.

**Comment by JonnyHaystack** [comment]
Right, I get you

**Comment by Nama** [comment]
I expanded some example script to make it work. The requirements are in the file. Only issue is that multiple windows of an application aren't handled.
https://gist.github.com/Nama/55786d0b2a8349d11f9013fa1a86e6b1

**Comment by MNolan147** [comment]
@Nama This script doesn't seem to work for me. It is able to save the tree, but can't restore it from the file.

**Comment by Nama** [comment]
Do you get any error? If yes, pls post.
Do you have more than one of the same monitor? If yes, tell me the names of them from `swaymsg -t get_outputs`.

**Comment by MNolan147** [comment]
@Nama No error, no duplicate monitors. When it finishes is says that the session was successfully loaded, but no programs were started and no already open programs were moved to the workspace they should be moved to.

**Comment by Nama** [comment]
The script doesn't start any programs.

Can you send the tree json the script saved?

**Comment by MNolan147** [comment]
{
    "id": 1,
    "type": "root",
    "orientation": "horizontal",
    "percent": null,
    "urgent": false,
    "marks": [],
    "focused": false,
    "layout": "splith",
    "border": "none",
    "current_border_width": 0,
    "rect": {
        "x": 0,
        "y": 0,
        "width": 1920,
        "height": 1200
    },
    "deco_rect": {
        "x": 0,
        "y": 0,
        "width": 0,
        "height": 0
    },
    "window_rect": {
        "x": 0,
        "y": 0,
        "width": 0,
        "height": 0
    },
    "geometry": {
        "x": 0,
        "y": 0,
        "width": 0,
        "height": 0
    },
    "name": "root",
    "window": null,
    "nodes": [
        {
            "id": 2147483647,
            "type": "output",
            "orientation": "horizontal",
            "percent": null,
            "urgent": false,
            "marks": [],
            "focused": false,
            "layout": "output",
            "border": "none",
            "current_border_width": 0,
            "rect": {
                "x": 0,
                "y": 0,
                "width": 1920,
                "height": 1200
            },
            "deco_rect": {
                "x": 0,
                "y": 0,
                "width": 0,
                "height": 0
            },
            "window_rect": {
                "x": 0,
                "y": 0,
                "width": 0,
                "height": 0
            },
            "geometry": {
                "x": 0,
                "y": 0,
                "width": 0,
                "height": 0
            },
            "name": "__i3",
            "window": null,
            "nodes": [
                {
                    "id": 2147483646,
                    "type": "workspace",
                    "orientation": "horizontal",
                    "percent": null,
                    "urgent": false,
                    "marks": [],
                    "focused": false,
                    "layout": "splith",
                    "border": "none",
                    "current_border_width": 0,
                    "rect": {
                        "x": 0,
                        "y": 0,
                        "width": 1920,
                        "height": 1200
                    },
                    "deco_rect": {
                        "x": 0,
                        "y": 0,
                        "width": 0,
                        "height": 0
                    },
                    "window_rect": {
                        "x": 0,
                        "y": 0,
                        "width": 0,
                        "height": 0
                    },
                    "geometry": {
                        "x": 0,
                        "y": 0,
                        "width": 0,
                        "height": 0
                    },
                    "name": "__i3_scratch",
                    "window": null,
                    "nodes": [],
                    "floating_nodes": [],
                    "focus": [],
                    "fullscreen_mode": 1,
                    "sticky": false
                }
            ],
            "floating_nodes": [],
            "focus": [
                2147483646
            ],
            "fullscreen_mode": 0,
            "sticky": false
        },
        {
            "id": 3,
            "type": "output",
            "orientation": "none",
            "percent": 1.0,
            "urgent": false,
            "marks": [],
            "focused": false,
            "layout": "output",
            "border": "none",
            "current_border_width": 0,
            "rect": {
                "x": 0,
                "y": 0,
                "width": 1920,
                "height": 1200
            },
            "deco_rect": {
                "x": 0,
                "y": 0,
                "width": 0,
                "height": 0
            },
            "window_rect": {
                "x": 0,
                "y": 0,
                "width": 0,
                "height": 0
            },
            "geometry": {
                "x": 0,
                "y": 0,
                "width": 0,
                "height": 0
            },
            "name": "eDP-1",
            "window": null,
            "nodes": [
                {
                    "id": 4,
                    "type": "workspace",
                    "orientation": "vertical",
                    "percent": null,
                    "urgent": false,
                    "marks": [],
                    "focused": false,
                    "layout": "splitv",
                    "border": "none",
                    "current_border_width": 0,
                    "rect": {
                        "x": 0,
                        "y": 27,
                        "width": 1920,
                        "height": 1173
                    },
                    "deco_rect": {
                        "x": 0,
                        "y": 0,
                        "width": 0,
                        "height": 0
                    },
                    "window_rect": {
                        "x": 0,
                        "y": 0,
                        "width": 0,
                        "height": 0
                    },
                    "geometry": {
                        "x": 0,
                        "y": 0,
                        "width": 0,
                        "height": 0
                    },
                    "name": "1:\uf269",
                    "window": null,
                    "nodes": [
                        {
                            "id": 9,
                            "type": "con",
                            "orientation": "horizontal",
                            "percent": 0.8158567774936062,
                            "urgent": false,
                            "marks": [],
                            "focused": false,
                            "layout": "splith",
                            "border": "none",
                            "current_border_width": 0,
                            "rect": {
                                "x": 0,
                                "y": 27,
                                "width": 1920,
                                "height": 957
                            },
                            "deco_rect": {
                                "x": 0,
                                "y": 0,
                                "width": 0,
                                "height": 0
                            },
                            "window_rect": {
                                "x": 0,
                                "y": 0,
                                "width": 0,
                                "height": 0
                            },
                            "geometry": {
                                "x": 0,
                                "y": 0,
                                "width": 0,
                                "height": 0
                            },
                            "name": null,
                            "window": null,
                            "nodes": [
                                {
                                    "id": 5,
                                    "type": "con",
                                    "orientation": "none",
                                    "percent": 0.765625,
                                    "urgent": false,
                                    "marks": [],
                                    "focused": false,
                                    "layout": "none",
                                    "border": "pixel",
                                    "current_border_width": 1,
                                    "rect": {
                                        "x": 0,
                                        "y": 27,
                                        "width": 1470,
                                        "height": 957
                                    },
                                    "deco_rect": {
                                        "x": 0,
                                        "y": 0,
                                        "width": 0,
                                        "height": 0
                                    },
                                    "window_rect": {
                                        "x": 1,
                                        "y": 1,
                                        "width": 1468,
                                        "height": 955
                                    },
                                    "geometry": {
                                        "x": 0,
                                        "y": 0,
                                        "width": 1000,
                                        "height": 1085
                                    },
                                    "name": "Restore workspaces in sway to displays and move applications to saved workspaces \u2014 Mozilla Firefox",
                                    "window": null,
                                    "nodes": [],
                                    "floating_nodes": [],
                                    "focus": [],
                                    "fullscreen_mode": 0,
                                    "sticky": false,
                                    "pid": 3143,
                                    "app_id": "firefox",
                                    "visible": true,
                                    "max_render_time": 0,
                                    "shell": "xdg_shell",
                                    "inhibit_idle": false,
                                    "idle_inhibitors": {
                                        "user": "fullscreen",
                                        "application": "none"
                                    }
                                },
                                {
                                    "id": 8,
                                    "type": "con",
                                    "orientation": "none",
                                    "percent": 0.234375,
                                    "urgent": false,
                                    "marks": [],
                                    "focused": true,
                                    "layout": "none",
                                    "border": "pixel",
                                    "current_border_width": 1,
                                    "rect": {
                                        "x": 1470,
                                        "y": 27,
                                        "width": 450,
                                        "height": 957
                                    },
                                    "deco_rect": {
                                        "x": 0,
                                        "y": 0,
                                        "width": 0,
                                        "height": 0
                                    },
                                    "window_rect": {
                                        "x": 1,
                                        "y": 1,
                                        "width": 448,
                                        "height": 955
                                    },
                                    "geometry": {
                                        "x": 0,
                                        "y": 0,
                                        "width": 1920,
                                        "height": 1173
                                    },
                                    "name": "python3 ~/.local/bin/sway_workspaces.py save startup",
                                    "window": null,
                                    "nodes": [],
                                    "floating_nodes": [],
                                    "focus": [],
                                    "fullscreen_mode": 0,
                                    "sticky": false,
                                    "pid": 4330,
                                    "app_id": "kitty",
                                    "visible": true,
                                    "max_render_time": 0,
                                    "shell": "xdg_shell",
                                    "inhibit_idle": false,
                                    "idle_inhibitors": {
                                        "user": "fullscreen",
                                        "application": "none"
                                    }
                                }
                            ],
                            "floating_nodes": [],
                            "focus": [
                                8,
                                5
                            ],
                            "fullscreen_mode": 0,
                            "sticky": false
                        },
                        {
                            "id": 10,
                            "type": "con",
                            "orientation": "none",
                            "percent": 0.18414322250639387,
                            "urgent": false,
                            "marks": [],
                            "focused": false,
                            "layout": "none",
                            "border": "pixel",
                            "current_border_width": 1,
                            "rect": {
                                "x": 0,
                                "y": 984,
                                "width": 1920,
                                "height": 216
                            },
                            "deco_rect": {
                                "x": 0,
                                "y": 0,
                                "width": 0,
                                "height": 0
                            },
                            "window_rect": {
                                "x": 1,
                                "y": 1,
                                "width": 1918,
                                "height": 214
                            },
                            "geometry": {
                                "x": 0,
                                "y": 0,
                                "width": 1920,
                                "height": 1173
                            },
                            "name": "cava",
                            "window": null,
                            "nodes": [],
                            "floating_nodes": [],
                            "focus": [],
                            "fullscreen_mode": 0,
                            "sticky": false,
                            "pid": 4399,
                            "app_id": "kitty",
                            "visible": true,
                            "max_render_time": 0,
                            "shell": "xdg_shell",
                            "inhibit_idle": false,
                            "idle_inhibitors": {
                                "user": "fullscreen",
                                "application": "none"
                            }
                        }
                    ],
                    "floating_nodes": [],
                    "focus": [
                        9,
                        10
                    ],
                    "fullscreen_mode": 1,
                    "sticky": false,
                    "num": 1,
                    "output": "eDP-1",
                    "representation": "V[H[firefox kitty] kitty]"
                },
                {
                    "id": 11,
                    "type": "workspace",
                    "orientation": "vertical",
                    "percent": null,
                    "urgent": false,
                    "marks": [],
                    "focused": false,
                    "layout": "splitv",
                    "border": "none",
                    "current_border_width": 0,
                    "rect": {
                        "x": 0,
                        "y": 27,
                        "width": 1920,
                        "height": 1173
                    },
                    "deco_rect": {
                        "x": 0,
                        "y": 0,
                        "width": 0,
                        "height": 0
                    },
                    "window_rect": {
                        "x": 0,
                        "y": 0,
                        "width": 0,
                        "height": 0
                    },
                    "geometry": {
                        "x": 0,
                        "y": 0,
                        "width": 0,
                        "height": 0
                    },
                    "name": "3:\uf1bc",
                    "window": null,
                    "nodes": [
                        {
                            "id": 20,
                            "type": "con",
                            "orientation": "horizontal",
                            "percent": 0.8158567774936062,
                            "urgent": false,
                            "marks": [],
                            "focused": false,
                            "layout": "splith",
                            "border": "none",
                            "current_border_width": 0,
                            "rect": {
                                "x": 0,
                                "y": 27,
                                "width": 1920,
                                "height": 957
                            },
                            "deco_rect": {
                                "x": 0,
                                "y": 0,
                                "width": 0,
                                "height": 0
                            },
                            "window_rect": {
                                "x": 0,
                                "y": 0,
                                "width": 0,
                                "height": 0
                            },
                            "geometry": {
                                "x": 0,
                                "y": 0,
                                "width": 0,
                                "height": 0
                            },
                            "name": null,
                            "window": null,
                            "nodes": [
                                {
                                    "id": 15,
                                    "type": "con",
                                    "orientation": "none",
                                    "percent": 0.765625,
                                    "urgent": false,
                                    "marks": [],
                                    "focused": false,
                                    "layout": "none",
                                    "border": "pixel",
                                    "current_border_width": 1,
                                    "rect": {
                                        "x": 0,
                                        "y": 27,
                                        "width": 1470,
                                        "height": 957
                                    },
                                    "deco_rect": {
                                        "x": 0,
                                        "y": 0,
                                        "width": 0,
                                        "height": 0
                                    },
                                    "window_rect": {
                                        "x": 1,
                                        "y": 1,
                                        "width": 1468,
                                        "height": 955
                                    },
                                    "geometry": {
                                        "x": 0,
                                        "y": 0,
                                        "width": 1918,
                                        "height": 389
                                    },
                                    "name": "ncspot",
                                    "window": null,
                                    "nodes": [],
                                    "floating_nodes": [],
                                    "focus": [],
                                    "fullscreen_mode": 0,
                                    "sticky": false,
                                    "pid": 4610,
                                    "app_id": "kitty",
                                    "visible": false,
                                    "max_render_time": 0,
                                    "shell": "xdg_shell",
                                    "inhibit_idle": false,
                                    "idle_inhibitors": {
                                        "user": "fullscreen",
                                        "application": "none"
                                    }
                                },
                                {
                                    "id": 18,
                                    "type": "con",
                                    "orientation": "vertical",
                                    "percent": 0.234375,
                                    "urgent": false,
                                    "marks": [],
                                    "focused": false,
                                    "layout": "splitv",
                                    "border": "none",
                                    "current_border_width": 0,
                                    "rect": {
                                        "x": 1470,
                                        "y": 27,
                                        "width": 450,
                                        "height": 957
                                    },
                                    "deco_rect": {
                                        "x": 0,
                                        "y": 0,
                                        "width": 0,
                                        "height": 0
                                    },
                                    "window_rect": {
                                        "x": 0,
                                        "y": 0,
                                        "width": 0,
                                        "height": 0
                                    },
                                    "geometry": {
                                        "x": 0,
                                        "y": 0,
                                        "width": 0,
                                        "height": 0
                                    },
                                    "name": null,
                                    "window": null,
                                    "nodes": [
                                        {
                                            "id": 17,
                                            "type": "con",
                                            "orientation": "none",
                                            "percent": 0.5005224660397074,
                                            "urgent": false,
                                            "marks": [],
                                            "focused": false,
                                            "layout": "none",
                                            "border": "pixel",
                                            "current_border_width": 1,
                                            "rect": {
                                                "x": 1470,
                                                "y": 27,
                                                "width": 450,
                                                "height": 479
                                            },
                                            "deco_rect": {
                                                "x": 0,
                                                "y": 0,
                                                "width": 0,
                                                "height": 0
                                            },
                                            "window_rect": {
                                                "x": 1,
                                                "y": 1,
                                                "width": 448,
                                                "height": 477
                                            },
                                            "geometry": {
                                                "x": 0,
                                                "y": 0,
                                                "width": 1918,
                                                "height": 584
                                            },
                                            "name": "cmatrix",
                                            "window": null,
                                            "nodes": [],
                                            "floating_nodes": [],
                                            "focus": [],
                                            "fullscreen_mode": 0,
                                            "sticky": false,
                                            "pid": 4708,
                                            "app_id": "kitty",
                                            "visible": false,
                                            "max_render_time": 0,
                                            "shell": "xdg_shell",
                                            "inhibit_idle": false,
                                            "idle_inhibitors": {
                                                "user": "fullscreen",
                                                "application": "none"
                                            }
                                        },
                                        {
                                            "id": 19,
                                            "type": "con",
                                            "orientation": "none",
                                            "percent": 0.4994775339602926,
                                            "urgent": false,
                                            "marks": [],
                                            "focused": false,
                                            "layout": "none",
                                            "border": "pixel",
                                            "current_border_width": 1,
                                            "rect": {
                                                "x": 1470,
                                                "y": 506,
                                                "width": 450,
                                                "height": 478
                                            },
                                            "deco_rect": {
                                                "x": 0,
                                                "y": 0,
                                                "width": 0,
                                                "height": 0
                                            },
                                            "window_rect": {
                                                "x": 1,
                                                "y": 1,
                                                "width": 448,
                                                "height": 476
                                            },
                                            "geometry": {
                                                "x": 0,
                                                "y": 0,
                                                "width": 1918,
                                                "height": 584
                                            },
                                            "name": "cbonsaiLooped.sh",
                                            "window": null,
                                            "nodes": [],
                                            "floating_nodes": [],
                                            "focus": [],
                                            "fullscreen_mode": 0,
                                            "sticky": false,
                                            "pid": 4753,
                                            "app_id": "kitty",
                                            "visible": false,
                                            "max_render_time": 0,
                                            "shell": "xdg_shell",
                                            "inhibit_idle": false,
                                            "idle_inhibitors": {
                                                "user": "fullscreen",
                                                "application": "none"
                                            }
                                        }
                                    ],
                                    "floating_nodes": [],
                                    "focus": [
                                        17,
                                        19
                                    ],
                                    "fullscreen_mode": 0,
                                    "sticky": false
                                }
                            ],
                            "floating_nodes": [],
                            "focus": [
                                15,
                                18
                            ],
                            "fullscreen_mode": 0,
                            "sticky": false
                        },
                        {
                            "id": 21,
                            "type": "con",
                            "orientation": "none",
                            "percent": 0.18414322250639387,
                            "urgent": false,
                            "marks": [],
                            "focused": false,
                            "layout": "none",
                            "border": "pixel",
                            "current_border_width": 1,
                            "rect": {
                                "x": 0,
                                "y": 984,
                                "width": 1920,
                                "height": 216
                            },
                            "deco_rect": {
                                "x": 0,
                                "y": 0,
                                "width": 0,
                                "height": 0
                            },
                            "window_rect": {
                                "x": 1,
                                "y": 1,
                                "width": 1918,
                                "height": 214
                            },
                            "geometry": {
                                "x": 0,
                                "y": 0,
                                "width": 1918,
                                "height": 584
                            },
                            "name": "cava",
                            "window": null,
                            "nodes": [],
                            "floating_nodes": [],
                            "focus": [],
                            "fullscreen_mode": 0,
                            "sticky": false,
                            "pid": 4798,
                            "app_id": "kitty",
                            "visible": false,
                            "max_render_time": 0,
                            "shell": "xdg_shell",
                            "inhibit_idle": false,
                            "idle_inhibitors": {
                                "user": "fullscreen",
                                "application": "none"
                            }
                        }
                    ],
                    "floating_nodes": [],
                    "focus": [
                        20,
                        21
                    ],
                    "fullscreen_mode": 1,
                    "sticky": false,
                    "num": 3,
                    "output": "eDP-1",
                    "representation": "V[H[kitty V[kitty kitty]] kitty]"
                },
                {
                    "id": 7,
                    "type": "workspace",
                    "orientation": "vertical",
                    "percent": null,
                    "urgent": false,
                    "marks": [],
                    "focused": false,
                    "layout": "splitv",
                    "border": "none",
                    "current_border_width": 0,
                    "rect": {
                        "x": 0,
                        "y": 27,
                        "width": 1920,
                        "height": 1173
                    },
                    "deco_rect": {
                        "x": 0,
                        "y": 0,
                        "width": 0,
                        "height": 0
                    },
                    "window_rect": {
                        "x": 0,
                        "y": 0,
                        "width": 0,
                        "height": 0
                    },
                    "geometry": {
                        "x": 0,
                        "y": 0,
                        "width": 0,
                        "height": 0
                    },
                    "name": "4:@",
                    "window": null,
                    "nodes": [
                        {
                            "id": 6,
                            "type": "con",
                            "orientation": "none",
                            "percent": 0.8158567774936062,
                            "urgent": false,
                            "marks": [],
                            "focused": false,
                            "layout": "none",
                            "border": "pixel",
                            "current_border_width": 1,
                            "rect": {
                                "x": 0,
                                "y": 27,
                                "width": 1920,
                                "height": 957
                            },
                            "deco_rect": {
                                "x": 0,
                                "y": 0,
                                "width": 0,
                                "height": 0
                            },
                            "window_rect": {
                                "x": 1,
                                "y": 1,
                                "width": 1918,
                                "height": 955
                            },
                            "geometry": {
                                "x": 0,
                                "y": 0,
                                "width": 960,
                                "height": 879
                            },
                            "name": "Inbox - School Email - Mozilla Thunderbird",
                            "window": null,
                            "nodes": [],
                            "floating_nodes": [],
                            "focus": [],
                            "fullscreen_mode": 0,
                            "sticky": false,
                            "pid": 3148,
                            "app_id": "thunderbird",
                            "visible": false,
                            "max_render_time": 0,
                            "shell": "xdg_shell",
                            "inhibit_idle": false,
                            "idle_inhibitors": {
                                "user": "fullscreen",
                                "application": "none"
                            }
                        },
                        {
                            "id": 22,
                            "type": "con",
                            "orientation": "none",
                            "percent": 0.18414322250639387,
                            "urgent": false,
                            "marks": [],
                            "focused": false,
                            "layout": "none",
                            "border": "pixel",
                            "current_border_width": 1,
                            "rect": {
                                "x": 0,
                                "y": 984,
                                "width": 1920,
                                "height": 216
                            },
                            "deco_rect": {
                                "x": 0,
                                "y": 0,
                                "width": 0,
                                "height": 0
                            },
                            "window_rect": {
                                "x": 1,
                                "y": 1,
                                "width": 1918,
                                "height": 214
                            },
                            "geometry": {
                                "x": 0,
                                "y": 0,
                                "width": 1918,
                                "height": 584
                            },
                            "name": "cava",
                            "window": null,
                            "nodes": [],
                            "floating_nodes": [],
                            "focus": [],
                            "fullscreen_mode": 0,
                            "sticky": false,
                            "pid": 4845,
                            "app_id": "kitty",
                            "visible": false,
                            "max_render_time": 0,
                            "shell": "xdg_shell",
                            "inhibit_idle": false,
                            "idle_inhibitors": {
                                "user": "fullscreen",
                                "application": "none"
                            }
                        }
                    ],
                    "floating_nodes": [],
                    "focus": [
                        22,
                        6
                    ],
                    "fullscreen_mode": 1,
                    "sticky": false,
                    "num": 4,
                    "output": "eDP-1",
                    "representation": "V[thunderbird kitty]"
                }
            ],
            "floating_nodes": [],
            "focus": [
                4,
                11,
                7
            ],
            "fullscreen_mode": 0,
            "sticky": false,
            "active": true,
            "dpms": true,
            "primary": false,
            "make": "Unknown",
            "model": "0xF392",
            "serial": "0x00000000",
            "scale": 1.0,
            "scale_filter": "nearest",
            "transform": "normal",
            "adaptive_sync_status": "disabled",
            "current_workspace": "1:\uf269",
            "modes": [
                {
                    "width": 1920,
                    "height": 1200,
                    "refresh": 60026
                },
                {
                    "width": 1920,
                    "height": 1200,
                    "refresh": 40019
                }
            ],
            "current_mode": {
                "width": 1920,
                "height": 1200,
                "refresh": 60026
            },
            "max_render_time": 0
        }
    ],
    "floating_nodes": [],
    "focus": [
        3
    ],
    "fullscreen_mode": 0,
    "sticky": false
}

**Comment by Nama** [comment]
Please send such long text as file. You should edit and delete your json tree now.
My script makes use of the `window_properties`, which your tree doesn't have once.
```json
              "window_properties": {
                "class": "firefox",
                "instance": "Navigator",
                "title": "Variable refresh rate - ArchWiki — Mozilla Firefox",
                "transient_for": null,
                "window_role": "browser",
                "window_type": "normal"
              }
```
I don't know why. The only differences seem to be that `window_properties` are missing and (probably for the same reason) `"window"` is always `null`. My applications have all an ID as the value.
Even tho, `"visible"` says `true`, these values are missing for you. That you (somehow) hid the windows would have been my first guess.

Do you have any workspaces which aren't vertically aligned? I have only one vertical workspace, but the `window_properties` aren't missing there. For me, only Sublime Text is missing these values.

My version of sway is 1:1.7.

**Comment by MNolan147** [comment]
Apologies for the long text. I've attached the file to this message.
I've run `swaymsg -t get_tree` and the `window_properties` don't appear there either

Workspace Tree:
[workspace_startup_tree.txt](https://github.com/swaywm/sway/files/9570420/workspace_startup_tree.txt)

**Comment by Nama** [comment]
Figured it out.
Only applications using xwayland get the `"window_properties"` data.
And somehow, ever of my apps do this except Sublime Text. (Firefox needs even a env var)

My script is useless.

**Comment by MNolan147** [comment]
You could change the script to also check for the `app_id` property, which seems to perform the same function as `window_propeties["class"]`.

**Comment by MNolan147** [comment]
And to move multiple windows of the same type, you could use the `id` property to distinguish between the different windows i.e., if the layout has 2 terminals that it needs to move to specific workstations, it loops through open terminals and moves them based on their `id`.

**Comment by Nama** [comment]
Oh yeah, `app_id` seems useful, thx.

Isn't the `id` of the node always another one, so we can't compare them?
Or do you mean like, keeping track of, which `id` was moved already and skipping to never moved ones?

**Comment by Nama** [comment]
Wayland apps are now working.
You need i3ipc from Github, newest commit. Last release is lacking that feature -.-
Also made dunstify and sh optional, without the need to edit the script.

I'll look into handling multiple windows now.
https://gist.github.com/Nama/55786d0b2a8349d11f9013fa1a86e6b1

**Comment by roselandgoose** [comment]
I've modified @Nama's script above to support multiple windows of the same program - for my many firefox windows...
I've also made the layout file argument optional.

https://git.sr.ht/~roselandgoose/Sway_Layout_Restore/blob/main/sway_workspaces.py

**Comment by Nama** [comment]
Oh that's nice! Thank you, @roselandgoose. I created a repo without your changes, so you can PR your changes. Would be more convenient to work on it in a repo (and move the discussion from this issue XD).
https://github.com/Nama/sway-workspaces

**Comment by JoshElias** [comment]
As a TLDR; for people looking for this feature, the sway dev doesn't want to add it but Nama has created a [plugin](https://github.com/Nama/swayrst) that somewhat recreates **i3-save-tree**.

**Comment by mishurov** [comment]
@Nama I can traverse recursively a saved json from `GET_TREE` and launch apps using `app_id`, or map `app_id` to an executable, instead of moving existing apps. I wrote a script for my personal use with my own unbloated sway ipc but If you're interested, you might take a look how I do it.
https://github.com/mishurov/applets/blob/master/sway_restore_workspace/srws.py
```sh
python3 ./srws.py --workspace myworkspacename --save ./test.json
python3 ./srws.py --load ./test.json
```

**Comment by jayson-garrison** [comment]
Adding support for the i3 equivalent for saving/restoring layouts for Sway should be reconsidered. It is already clear from this thread that the implementation for this on the Sway side should be different than that of i3. It is also apparent that writing scripts to do this is not ideal and ad-hoc. Saving configurations of application tile layouts, I do not believe, is a fringe feature. Let me give you my use case in my industry:

We are delivering a software/hardware suite to a customer for a certain task. This task involves a lot of hardware that is interfaced to the user using our software suite. This takes the form of GUIs presenting hardware information like statuses, graphs, maps, and other specific capabilities implemented by us using the hardware supplied by us. The user also wants to use third party applications such as a text editor, or Adobe, or terminals that are without our software suite.

The customer will provide users that will actually use our system to perform the task. What we need is a window manager that can tile all of our software applications and their desired third party applications across multiple displays which will then be presented to the user. For this task, we need specific configurable and savable layouts that the users will be trained on. Having known configurations is critical for this use case especially when the users are operating the system in a real-time setting. It is also important because multiple users must be trained to use the system and we have to write the documentation for training, a specific layout of all the GUIs will be referenced in training. We can make our own proprietary Wayland compositor but Sway is a good option--if it supported configure/restore layouts.

This use case is not niche. Imagine any situation where a team of software developers have to deliver a product to a customer where the users will interact with the hardware/software suite via many GUIs. Consider the fact that the customer and their users, in a significant portion of these cases, have little to no software knowledge and cannot be expected to work with scripts to do this.
Here are examples I can think of off the top of my head to illustrate this: music company that wants a comprehensive sound engineering software suite, refinery company that wants a software solution that can monitor/control all of their machinery, medical professionals that want a suite of GUIs to interface with their machines, basically any situation where there is much hardware and software that is used to interface with it. It would be nice to use Sway in these situations.

Would love to hear anyone's feedback. Thanks.

**Comment by sigmavirus24** [comment]
@jayson-garrison just a sway user following the discussion here, but if your company needs this, they should develop it with the maintainers to be up to whatever standards they set and contribute it.

If I were you, or your employer, I would not show up with a use case I am willing to bet no one else here cares about and whinge about a free project made in people's personal time for their own personal goals not meeting your company's product needs. 

I'm also a strong advocate for a project and its maintainers saying No until they're blue in the face to preserve their sanity and whatever priorities they have for the pilot. I'm also, interested in the feature but now want them to continue to reject it just to spite your employer

**Comment by 0xRamsi** [comment]
I wanted this feature as well, but found a good alternative with writing a little script. One use case was that I wanted the same setup to learn Rust. Here is the solution:

```
## some other setup possible here.
swaymsg workspace 1
swaymsg exec -- gnome-terminal --working-directory=/home/user/dev/rust/the-book/
sleep 0.4   ## This is a bit hacky, but it works.
swaymsg workspace 2
swaymsg exec -- /usr/bin/firefox -P "Rust"
swaymsg exec /usr/bin/codium
swaymsg exec /usr/bin/vlc
## here I connect to my bluetooth speaker and such.
```

I saved that into a file called "rust" in a folder that is in my path (don't forget the `chmod +x rust`), and now I can just launch dmenu, and from there call "rust". It will setup everything.
Actually, this has the advantage of shell commands. E.g., I could already load a folder into codium, or a playlist into VLC.

I'm unfamiliar with the i3 version, probably some nuances will be different, but this is good for me. Feel free to post your improvements, after all we're here for the beauty of open source. :)
